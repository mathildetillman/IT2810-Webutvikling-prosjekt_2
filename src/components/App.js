import React, { useState, useEffect } from "react";
import RadioButtons from "./Radiobuttons";
import ArtContainer from "./ArtContainer";

const ART_CATEGORIES = [
  {
    category: "audio",
    choices: ["acoustic", "cinematic", "rock"],
  },
  {
    category: "image",
    choices: ["butterfly", "flower", "planet"],
  },
  {
    category: "text",
    choices: ["lyric", "poem", "quote"],
  },
];

export default function App() {
  //Keeps track of active categories:
  const [activeAudioCategory, setActiveAudioCategory] = useState("acoustic");
  const [activeImageCategory, setActiveImageCategory] = useState("butterfly");
  const [activeTextCategory, setActiveTextCategory] = useState("lyric");

  //Stores media being displayed:
  const [activeTab, setActiveTab] = useState("0");
  const [text, setText] = useState(null);
  const [image, setImage] = useState(null);
  const [audio, setAudio] = useState(null);

  function changeActiveArt(category, active) {
    switch (category) {
      case "audio":
        setActiveAudioCategory(active);
        break;
      case "image":
        setActiveImageCategory(active);
        break;
      case "text":
        setActiveTextCategory(active);
        break;
      default:
        break;
    }
  }

  function changeActiveTab(active) {
    setActiveTab(active);
  }

  //Fetch texts
  useEffect(() => {
    // Displays nothing while fetching new text to enhance UX
    setText(null);

    const url =
      process.env.PUBLIC_URL + "/content/text/" + activeTextCategory + ".json";
    let data = sessionStorage.getItem(url);

    //If text stored in sessionStorage, no need to fetch again
    if (data) {
      setText(JSON.parse(data).items[activeTab]);
    }

    //If not fetched before; fetch and store in sessionStorage
    else {
      fetch(url)
        .then((res) => res.json())
        .then((res) => {
          setText(res.items[activeTab]);
          sessionStorage.setItem(url, JSON.stringify(res));
        });
    }
  }, [activeTextCategory, activeTab]);

  //Fetch audio
  useEffect(() => {
    // Displays nothing while fetching new audio to enhance UX
    setAudio(null);

    const url =
      process.env.PUBLIC_URL +
      `/content/audio/${activeAudioCategory}/${activeAudioCategory}${activeTab}.mp3`;
    let data = sessionStorage.getItem(url);

    //If audio stored in sessionStorage, no need to fetch again
    if (data) {
      setAudio(JSON.parse(data));
    }

    //If not fetched before; fetch and store in sessionStorage
    else {
      setAudio(url);
      sessionStorage.setItem(url, JSON.stringify(url));
    }
  }, [activeAudioCategory, activeTab]);

  //Fetch image
  useEffect(() => {
    // Displays nothing while fetching new image to enhance UX
    setImage(null);

    const url =
      process.env.PUBLIC_URL +
      `/content/image/${activeImageCategory}/${activeImageCategory}${activeTab}.svg`;
    let data = sessionStorage.getItem(url);

    //If image stored in sessionStorage, no need to fetch again
    if (data) {
      setImage(JSON.parse(data));
    }

    //If not fetched before; fetch and store in sessionStorage
    else {
      fetch(url)
        .then((res) => res.text())
        .then((svg) => {
          setImage(svg);
          sessionStorage.setItem(url, JSON.stringify(svg));
        });
    }
  }, [activeImageCategory, activeTab]);

  //Store favourite artcollection in localStorage
  function makeFavourite() {
    if (typeof Storage !== "undefined") {
      localStorage.setItem("favTab", activeTab);
      localStorage.setItem("favImage", activeImageCategory);
      localStorage.setItem("favAudio", activeAudioCategory);
      localStorage.setItem("favText", activeTextCategory);
    }
  }

  //Get favourite artcollection from localStorage (if it exists)
  function showFavourite(e) {
    e.preventDefault();
    if (typeof Storage !== "undefined") {
      if (localStorage.favImage) {
        setActiveAudioCategory(localStorage.getItem("favAudio"));
        setActiveImageCategory(localStorage.getItem("favImage"));
        setActiveTextCategory(localStorage.getItem("favText"));
        setActiveTab(localStorage.getItem("favTab"));
      } else {
        console.log("You have no favourite :(");
      }
    }
  }

  return (
    <div>
      <div className="header">
        <h1>ART EXHIBITION</h1>
      </div>

      <div className="main">
        <ArtContainer
          activeTab={activeTab}
          image={image}
          text={text}
          audio={audio}
          changeActiveTab={changeActiveTab}
        />
        <div className="btn-div">
          <button className="fav-btn" onClick={makeFavourite}>
            Make Favourite
          </button>
          <button className="fav-btn" onClick={showFavourite}>
            Show Favourite
          </button>
        </div>
      </div>

      <div>
        <div className="flex-container-radio footer">
          {ART_CATEGORIES.map((element, index) => {
            return (
              <RadioButtons
                changeActiveArt={changeActiveArt}
                active={[
                  activeAudioCategory,
                  activeImageCategory,
                  activeTextCategory,
                ]}
                key={index}
                category={element.category}
                choices={element.choices}
              />
            );
          })}
        </div>
      </div>
    </div>
  );
}
