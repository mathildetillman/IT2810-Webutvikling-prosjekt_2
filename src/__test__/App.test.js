import React from "react";
import renderer from "react-test-renderer";
import App from "../components/App";



it("Renders correctly", ()=>{
    const tree = renderer.create(<App/>).toJSON();
    expect(tree).toMatchSnapshot();

})