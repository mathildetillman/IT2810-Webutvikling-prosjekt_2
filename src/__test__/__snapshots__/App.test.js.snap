// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="grid-container"
>
  <div
    className="header"
  >
    <h1>
      ART EXHIBITION
    </h1>
  </div>
  <div
    className="main"
  >
    <div
      className="card"
    >
      <div>
        <div
          className="tab flex-container-tab"
        >
          <button
            className="active tab"
            name="0"
            onClick={[Function]}
          >
            Art 1
          </button>
          <button
            className="tab"
            name="1"
            onClick={[Function]}
          >
            Art 2
          </button>
          <button
            className="tab"
            name="2"
            onClick={[Function]}
          >
            Art 3
          </button>
          <button
            className="tab"
            name="3"
            onClick={[Function]}
          >
            Art 4
          </button>
        </div>
      </div>
      <div
        className="grid-container-art"
      >
        <div
          className="grid-item-image"
          dangerouslySetInnerHTML={
            Object {
              "__html": null,
            }
          }
        />
        <div
          className="grid-item-text"
        />
        <div
          className="grid-item-audio"
        >
          <audio
            controls={true}
            src={null}
            type="audio/mp3"
          />
        </div>
      </div>
    </div>
    <div
      className="btn-div"
    >
      <button
        className="fav-btn"
        onClick={[Function]}
      >
        Make Favourite
      </button>
      <button
        className="fav-btn"
        onClick={[Function]}
      >
        Show Favourite
      </button>
    </div>
  </div>
  <div>
    <div
      className="flex-container-radio footer"
    >
      <form>
        <h1>
          audio
        </h1>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={true}
              className="pointer"
              name="audio"
              onChange={[Function]}
              type="radio"
              value="acoustic"
            />
            acoustic
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="audio"
              onChange={[Function]}
              type="radio"
              value="cinematic"
            />
            cinematic
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="audio"
              onChange={[Function]}
              type="radio"
              value="rock"
            />
            rock
            <span
              className="checkmark"
            />
          </label>
        </div>
      </form>
      <form>
        <h1>
          image
        </h1>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={true}
              className="pointer"
              name="image"
              onChange={[Function]}
              type="radio"
              value="butterfly"
            />
            butterfly
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="image"
              onChange={[Function]}
              type="radio"
              value="flower"
            />
            flower
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="image"
              onChange={[Function]}
              type="radio"
              value="planet"
            />
            planet
            <span
              className="checkmark"
            />
          </label>
        </div>
      </form>
      <form>
        <h1>
          text
        </h1>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={true}
              className="pointer"
              name="text"
              onChange={[Function]}
              type="radio"
              value="lyric"
            />
            lyric
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="text"
              onChange={[Function]}
              type="radio"
              value="poem"
            />
            poem
            <span
              className="checkmark"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label
            className="pointer"
          >
            <input
              checked={false}
              className="pointer"
              name="text"
              onChange={[Function]}
              type="radio"
              value="quote"
            />
            quote
            <span
              className="checkmark"
            />
          </label>
        </div>
      </form>
    </div>
  </div>
</div>
`;
